src/main/java/org/openstreetmap/atlas/checks/validation/relations/InvalidTurnRestrictionCheck.java

```diff
@@ -129,4 +193,173 @@ private String getInstructionFromInvalidReason(final String invalidReason)
+    private boolean isUTurn(final Angle angle)
+    {
+        return Math.abs(angle.asDegrees()) > this.uturnAngleThreshold.asDegrees();
+    }
+
+    /**
+     * Return true if the U_TURN restriction doesn't match the topology
+     *
+     * @param TurnRestrictionTag turnRestrictionTag
+     * 
+     * @param Angle turnAngle
+     *
+     * @return true if the U_TURN restriction doesn't match the topology
+     */
+    private boolean isUTurnTopologyViolated(final TurnRestrictionTag turnRestrictionTag, final Angle turnAngle)
+    {
+        return TurnRestrictionTag.NO_U_TURN == turnRestrictionTag 
+                && !this.isUTurn(turnAngle);
+    }    
+
+    /**
+     * Return true if the turn restriction tag satisfies the topology
+     *
+     * @param Relation relation
+     *
+     * @return true if the turn restriction tag satisfies the topology
+     */
+    private boolean isValidTopology(final Relation relation)
+    {
+        // Build the via members
+        final Set<AtlasItem> viaMembers = relation.members().stream()
+                .filter(member -> member.getRole().equals(RelationTypeTag.RESTRICTION_ROLE_VIA))
+                .filter(member -> member.getEntity() instanceof Node
+                        || member.getEntity() instanceof Edge)
+                .map(RelationMember::getEntity).map(entity -> (AtlasItem) entity)
+                .collect(Collectors.toSet());
+
+        // Filter the members to extract only the "from" members that are connected at the end
+        // to VIA members.
+        final Set<Edge> fromMembers = new TreeSet<>();
+        relation.members().stream()
+                .filter(member -> member.getRole().equals(RelationTypeTag.RESTRICTION_ROLE_FROM)
+                        && member.getEntity() instanceof Edge
+                        && (!viaMembers.isEmpty()
+                                && ((Edge) member.getEntity()).isConnectedAtEndTo(viaMembers)))
+                .forEach(member -> fromMembers.add((Edge) member.getEntity()));
+
+        // Filter the members to extract only the "TO" members that are connected at the
+        // beginning to VIA members
+        final Set<Edge> toMembers = new TreeSet<>();
+        relation.members().stream()
+                .filter(member -> member.getRole().equals(RelationTypeTag.RESTRICTION_ROLE_TO)
+                        && member.getEntity() instanceof Edge
+                        && (!viaMembers.isEmpty()
+                                && ((Edge) member.getEntity()).isConnectedAtStartTo(viaMembers)))
+                .forEach(member -> toMembers.add((Edge) member.getEntity()));
+
+        // find the turnAngle between TO and FROM using angle subtraction
+        // the resulting turnAngle will be between -180 to 180 degrees
+        // 
+        //    from
+        //      \
+        //       \
+        //        \------> to
+        //
+
+        final Angle turnAngle = toMembers.iterator().next().asPolyLine().initialHeading().get()
```

Comment:
You should be able use the TurnRestriction created in the flag method to derive these headings (e.g. `turnRestriction.getTo().initialHeading()`). That will save you finding the to, via, and from members yourself.

Reply:
Thanks for your feedback! I will clean this up and push out the final PR when ready! Thanks again for reviewing!